File: src/features/quran/data/quranRepository.ts
/**
 * Quran repository — offline-first.
 * Base dataset: quran-json (Tanzil Uthmani text + Sahih International translation),
 * code-split per surah and cached in IndexedDB. The live API (alquran.cloud) enriches
 * ayahs with juz/page/tafsir and more editions when the network allows.
 */
import { db } from '@core/db/localdb';
import { toAppError } from '@core/errors/errors';
import surahIndex from 'quran-json/dist/chapters/en/index.json';
import type {
  AyahRange,
  CachedAyah,
  CachedSurah,
  QuranSearchResult,
  Reciter,
} from '../domain/types';
import { SAHIH_INTL_SOURCE, TANZIL_SOURCE } from '../domain/types';
import { normalizeArabic, parseAyahReference } from '../domain/arabic';

interface BundledSurahIndexEntry {
  id: number;
  name: string;
  transliteration: string;
  type: string;
  total_verses: number;
}

interface BundledVerse {
  id: number;
  text: string;
  translation?: string;
}

interface BundledChapter extends BundledSurahIndexEntry {
  verses: BundledVerse[];
}

/** Lazy per-surah chunks — Vite code-splits each chapter JSON. */
const chapterModules = import.meta.glob<BundledChapter>(
  '/node_modules/quran-json/dist/chapters/en/[0-9]*.json',
  { import: 'default' },
);

function chapterLoader(surah: number): (() => Promise<BundledChapter>) | null {
  const path = `/node_modules/quran-json/dist/chapters/en/${surah}.json`;
  return chapterModules[path] ?? null;
}

export function listSurahs(): CachedSurah[] {
  return (surahIndex as BundledSurahIndexEntry[]).map((s) => ({
    number: s.id,
    name: s.name,
    transliteration: s.transliteration,
    revelation: s.type === 'meccan' ? 'meccan' : 'medinan',
    ayahCount: s.total_verses,
  }));
}

const SURAHS = listSurahs();

export function surahByNumber(n: number): CachedSurah | null {
  return SURAHS.find((s) => s.number === n) ?? null;
}

function toCachedAyah(surah: number, verse: BundledVerse): CachedAyah {
  return {
    key: `${surah}:${verse.id}`,
    surah,
    ayah: verse.id,
    text: verse.text,
    translation: verse.translation ?? null,
    translationLang: 'en',
    juz: null,
    page: null,
    hizb: null,
    tafsir: null,
    source: TANZIL_SOURCE,
  };
}

/** Load (and cache) all ayahs of one surah. */
export async function getSurahAyahs(surah: number): Promise<CachedAyah[]> {
  const cached = await db.quranAyahs.where('surah').equals(surah).toArray();
  const meta = surahByNumber(surah);
  if (meta && cached.length === meta.ayahCount) {
    return cached.sort((a, b) => a.ayah - b.ayah);
  }
  const loader = chapterLoader(surah);
  if (!loader) throw toAppError(new Error(`Surah ${surah} not found`), 'validation');
  try {
    const chapter = await loader();
    const ayahs = chapter.verses.map((v) => toCachedAyah(surah, v));
    await db.quranAyahs.bulkPut(ayahs);
    return ayahs;
  } catch (error) {
    throw toAppError(error, 'storage');
  }
}

export async function getAyah(surah: number, ayah: number): Promise<CachedAyah | null> {
  const ayahs = await getSurahAyahs(surah);
  return ayahs.find((a) => a.ayah === ayah) ?? null;
}

export async function getAyahRange(range: AyahRange): Promise<CachedAyah[]> {
  const ayahs = await getSurahAyahs(range.surah);
  return ayahs.filter((a) => a.ayah >= range.from && a.ayah <= range.to);
}

/**
 * Search the whole Quran. A reference query ("2:255") resolves directly;
 * otherwise a normalized Arabic/translation substring search runs over all surahs
 * (chapters load lazily and are cached after the first search).
 */
export async function searchQuran(query: string, limit = 30): Promise<QuranSearchResult[]> {
  const trimmed = query.trim();
  if (trimmed.length === 0) return [];

  const ref = parseAyahReference(trimmed);
  if (ref) {
    const meta = surahByNumber(ref.surah);
    if (!meta || ref.ayah > meta.ayahCount) return [];
    const ayah = await getAyah(ref.surah, ref.ayah);
    return ayah ? [{ ayah, surahName: meta.name }] : [];
  }

  // Surah-name match → return the surah opening ayahs.
  const normQuery = normalizeArabic(trimmed);
  const results: QuranSearchResult[] = [];
  const nameMatch = SURAHS.find(
    (s) =>
      normalizeArabic(s.name) === normQuery ||
      s.transliteration.toLowerCase() === trimmed.toLowerCase(),
  );
  if (nameMatch) {
    const ayahs = await getSurahAyahs(nameMatch.number);
    for (const ayah of ayahs.slice(0, limit)) {
      results.push({ ayah, surahName: nameMatch.name });
    }
    return results;
  }

  if (normQuery.length < 3) return [];
  const lowerQuery = trimmed.toLowerCase();
  for (const surah of SURAHS) {
    const ayahs = await getSurahAyahs(surah.number);
    for (const ayah of ayahs) {
      const matchesArabic = normalizeArabic(ayah.text).includes(normQuery);
      const matchesTranslation =
        ayah.translation !== null && ayah.translation.toLowerCase().includes(lowerQuery);
      if (matchesArabic || matchesTranslation) {
        results.push({ ayah, surahName: surah.name });
        if (results.length >= limit) return results;
      }
    }
  }
  return results;
}

/** Translation source metadata (per-language registry, extensible). */
export function translationSource(): typeof SAHIH_INTL_SOURCE {
  return SAHIH_INTL_SOURCE;
}

/* ---------------- Audio ---------------- */

/** Cumulative ayah offsets to map (surah, ayah) → global ayah number (1..6236). */
const CUMULATIVE: number[] = (() => {
  const offsets: number[] = [0];
  for (const s of SURAHS) offsets.push(offsets[offsets.length - 1]! + s.ayahCount);
  return offsets;
})();

export function globalAyahNumber(surah: number, ayah: number): number {
  return (CUMULATIVE[surah - 1] ?? 0) + ayah;
}

/** Public CDN recitation audio (per-ayah MP3, streamed at play time). */
export function ayahAudioUrl(reciter: Reciter, surah: number, ayah: number): string {
  return `https://cdn.islamic.network/quran/audio/128/${reciter.edition}/${globalAyahNumber(surah, ayah)}.mp3`;
}

File: flutter_app/lib/features/quran/data/quran_repository.dart
/// Quran repository — bundled quran-json assets (same dataset as the PWA),
/// lazy per-surah loading with an in-memory cache, reference and normalized
/// Arabic search. Every ayah is served as a SOURCE_LOCK LockedText.
library;

import 'dart:convert';

import 'package:flutter/services.dart' show AssetBundle, rootBundle;

import '../../../core/arabic/arabic.dart';
import '../../../core/sourcelock/source_lock.dart';
import '../domain/models.dart';

class QuranRepository {
  QuranRepository({AssetBundle? bundle}) : _bundle = bundle ?? rootBundle;

  final AssetBundle _bundle;
  List<Surah>? _surahs;
  final Map<int, List<Ayah>> _chapters = {};

  Future<List<Surah>> listSurahs() async {
    if (_surahs != null) return _surahs!;
    final raw = await _bundle.loadString('assets/quran/index.json');
    final list = (jsonDecode(raw) as List)
        .cast<Map<String, dynamic>>()
        .map(Surah.fromJson)
        .toList(growable: false);
    _surahs = list;
    return list;
  }

  Future<Surah?> surahByNumber(int n) async =>
      (await listSurahs()).where((s) => s.number == n).firstOrNull;

  Future<List<Ayah>> getSurahAyahs(int surah) async {
    final cached = _chapters[surah];
    if (cached != null) return cached;
    final raw = await _bundle.loadString('assets/quran/en/$surah.json');
    final json = jsonDecode(raw) as Map<String, dynamic>;
    final verses = (json['verses'] as List).cast<Map<String, dynamic>>();
    final ayahs = verses
        .map(
          (v) => Ayah(
            surah: surah,
            ayah: v['id'] as int,
            locked: lockText(v['text'] as String, tanzilSource),
            translation: v['translation'] as String?,
          ),
        )
        .toList(growable: false);
    _chapters[surah] = ayahs;
    return ayahs;
  }

  Future<Ayah?> getAyah(int surah, int ayah) async {
    if (surah < 1 || surah > 114) return null;
    final ayahs = await getSurahAyahs(surah);
    return ayahs.where((a) => a.ayah == ayah).firstOrNull;
  }

  /// Reference search ("2:255") resolves directly; otherwise a normalized
  /// Arabic or English substring search over all surahs.
  Future<List<QuranSearchResult>> search(String query, {int limit = 20}) async {
    final trimmed = query.trim();
    if (trimmed.isEmpty) return const [];

    final surahs = await listSurahs();

    final ref = parseAyahReference(trimmed);
    if (ref != null) {
      final ayah = await getAyah(ref.surah, ref.ayah);
      if (ayah == null) return const [];
      final name = surahs.firstWhere((s) => s.number == ref.surah).name;
      return [QuranSearchResult(ayah, name)];
    }

    final needle = normalizeArabic(trimmed);
    final needleLower = trimmed.toLowerCase();
    final results = <QuranSearchResult>[];

    // Surah-name match first: return that surah's opening ayahs.
    final nameMatch = surahs.where(
      (s) =>
          normalizeArabic(s.name).contains(needle) ||
          s.transliteration.toLowerCase().contains(needleLower),
    );
    for (final s in nameMatch.take(1)) {
      final ayahs = await getSurahAyahs(s.number);
      for (final a in ayahs.take(limit)) {
        results.add(QuranSearchResult(a, s.name));
      }
      if (results.isNotEmpty) return results;
    }

    for (final s in surahs) {
      final ayahs = await getSurahAyahs(s.number);
      for (final a in ayahs) {
        final hit =
            normalizeArabic(a.text).contains(needle) ||
            (a.translation?.toLowerCase().contains(needleLower) ?? false);
        if (hit) {
          results.add(QuranSearchResult(a, s.name));
          if (results.length >= limit) return results;
        }
      }
    }
    return results;
  }

  /// Deterministic verse of the day from the verified dataset.
  Future<QuranSearchResult> verseOfDay(DateTime now) async {
    final surahs = await listSurahs();
    final dayIndex = now.difference(DateTime.utc(2024)).inDays % surahs.length;
    final surah = surahs[dayIndex < 0 ? 0 : dayIndex];
    final ayahs = await getSurahAyahs(surah.number);
    final ayah = ayahs[now.day % ayahs.length];
    return QuranSearchResult(ayah, surah.name);
  }
}
